### DISCUSSION TITLE
Add side value to BelongsToManyMultiSelect

### DISCUSSION THREAD

**Discussion by GregoireGaonach** [filamentphp/filament#1737] [Help]
Good evening,

I use `BelongsToManyMultiSelect` with a _Many To Many Polymorphic_ relationship and it works wonderfully well.

My pivot table looks like this one:
```txt
group_id | groupable_type  | groupable_id
---------+-----------------+--------------
1        | App\Models\User | 1
```

Now, let's say that I want to add a `role` column to my pivot table:
```txt
group_id | groupable_type  | groupable_id | role
---------+-----------------+--------------+----------
1        | App\Models\User | 1            | president
2        | App\Models\User | 1            | member
```

**The question is: how to add a value to `role` ?**

I though about two solutions:
1. First using a `MorphManyRepeater` but Filament only allows to use it with _On of many Polymorphic_ relationship.
2. Use 2 `BelongsToManyMultiSelect` with default role value, one for `president` and the other one for `member` by creating a custum field `BelongsToManyMultiSelectWithColumn` extending `BelongsToManyMultiSelect`. But it can become heavy in case of multiple roles.

The final code could looks lile this one:
```php
<?php
// ...
BelongsToManyMultiSelectWithColumn::make('groups')
    ->relationship('groups', 'name'),
    ->columns(['role' => 'president']),
BelongsToManyMultiSelectWithColumn::make('groups')
    ->relationship('groups', 'name'),
    ->columns(['role' => 'member']),
```

 * How would you manage to solve this problem?

**Comment by danharrin** [comment]
Someone should PR a `pivotData()` (instead of `columns()`) method that defines pivot data to be inserted when the `->saveRelationshipsUsing()` callback is executed and the relationships are attached. You've definitely got the right idea 👍

**Comment by GregoireGaonach** [comment]
Thank you very much for your answer @danharrin, it was very helpful. 

I created a method that works for this specific problem. I am not sure it is the best way to do it, but here it is. Now, it is time to understand the way Filament uses unit tests 😄 

```php
public function pivotData(array $pivotData)
{
    $this->saveRelationshipsUsing(function (BelongsToManyMultiSelect $component, ?array $state) use ($pivotData): void {
        $pivotKeyName = $component->getRelationship()->getRelatedPivotKeyName();
        $component->getRelationship()->sync(
            collect($state)
                ->map(function($item, $key) use ($pivotKeyName, $pivotData){
                        return $key = array_merge([$pivotKeyName => $item], $pivotData);
                    })
                ->toArray() ?? []
        );
    });
    return $this;
}
```

**Reply by danharrin** [reply]
Yeah thats kinda how we should do it, IMO we should store the pivot data in a property on the component, and then inside the original `saveRelationshipsUsing()` we can access this data. Instead of duplicating the syncing logic. Does that make sense?

Also, since some of our APIs are still in development, we don't have unit tests for specific components yet, but this is coming soon. So don't worry too much about those.

**Comment by GregoireGaonach** [comment]
Good afternoon,

Thanks again for your precious answer, everything is working well but I still have a problem that I can't figure out myself. 

Since my `BelongsToManyMultiSelect` field save its data with a specific pivot key data, how may I edit the `where` close?

For a _Many To Many Polymorphic_ relationship, I have the following condition. When I hardcode a role name, it works like a charm within Filament.

```php
/**
 * Get all of the users's groups
 *
 * @return \Illuminate\Database\Eloquent\Relations\morphToMany
 */
public function groups($role = null): morphToMany
{
    return $this->morphToMany(Group::class, 'groupable', 'group_users')->withPivot('role')->whereRole($role)->withTimestamps();
}
```

How may I define a value for the `$role` variable within the Filament logic?

Because from what I understood, there are only three [event in the lifecycle of a field](https://filamentphp.com/docs/2.x/forms/advanced) (`afterStateHydrated`, `afterStateUpdated`, `dehydrateStateUsing`) but they seems all to be called after the relationship. I tried to work with `options` but it seems to have no effect on the relationship.

Cheers mate and again thank you for the hard work.

**Reply by danharrin** [reply]
Hm, could you add the `->whereRole()` to the callback in the third param of `relationship()`? Which accepts `$query`.
